### PR TITLE
fix #2093 vmware compute resources don't support folders

### PR DESCRIPTION
**Pull request by mmatuska**
Fix issue [#2093](http://projects.theforeman.org/issues/2093)

traverse all datacenter folders and their subvolders
store vSphere folder path string in MiniServer object
add folder column to "Virtual Machines" view

### CODE REVIEW

**Comment by mmatuska** [comment]
The allbyfolder function may be moved to private if desired

File: app/services/fog_extensions/vsphere/mini_server.rb [hunk 1]
```diff
@@ -4,19 +4,24 @@ class MiniServer
...
         @memory   = raw.config.hardware.memoryMB * 1024 * 1024
         @ready    = raw.runtime.powerState == "poweredOn"
+        @path     = path
       end
 
       def state
         raw.runtime.powerState
       end
 
+      def path
```
Comment (domcleal) [inline app/services/fog_extensions/vsphere/mini_server.rb]: You can add this to attr_reader on line 4, which will automatically create a getter for `@path`.

File: app/services/fog_extensions/vsphere/mini_servers.rb [hunk 2]
```diff
@@ -10,13 +10,30 @@ def initialize client, dc
       end
 
       def all(filters = { })
-        dc.vmFolder.childEntity.grep(RbVmomi::VIM::VirtualMachine).map do |server|
-          MiniServer.new(server)
+        ret = []
+        allbyfolder(dc.vmFolder, nil).each do |entry|
+          ret.push MiniServer.new(entry[:vm], entry[:path])
```
Comment (domcleal) [inline app/services/fog_extensions/vsphere/mini_servers.rb]: The "map" method on arrays is useful here, it'll generate an output array from an input one, e.g.

```
def all(filters = {})
  allbyfolder(dc.vmFolder).map { |vm| MiniServer.new(vm[:vm], vm[:path]) }
end
```

File: app/services/fog_extensions/vsphere/mini_servers.rb [hunk 3]
```diff
@@ -10,13 +10,30 @@ def initialize client, dc
...
         end
+        ret
       end
 
+      def allbyfolder(folder, path = nil, recursing = false)
+        ret = []
+        if recursing
```
Comment (domcleal) [inline app/services/fog_extensions/vsphere/mini_servers.rb]: I don't think the recursing argument is needed here, you can just let path equal nil when you start to call the method and assume if it's non-nil then you're recursing.

**Comment by domcleal** [comment]
Thanks, this works really well.  Would you be able to look at the VM creation side too please?

There's a patch [back on the bug report](http://projects.theforeman.org/issues/2093#note-2) which adds the folder selection to the new VM form so you're able to create a VM within a specific folder.  Maybe you can adapt this?

**Comment by domcleal** [comment]
Ugh, ignore everything in the last comment, we've already got that.  More coffee required.

File: app/services/fog_extensions/vsphere/mini_servers.rb [hunk 4]
```diff
@@ -10,13 +10,28 @@ def initialize client, dc
...
       end
 
+      def allbyfolder(folder, path = nil)
+        ret = []
+        unless folder == @dc.vmFolder 
+          path = path.nil? ? folder.name : path + '/' + folder.name
+        end
+        folder.childEntity.each do |entity|
+          if entity.is_a?(RbVmomi::VIM::Folder)
+            ret = ret + (allbyfolder(entity, path))
```
Comment (ohadlevy) [inline app/services/fog_extensions/vsphere/mini_servers.rb:25]: won't this slow it down a lot? maybe its better to have a dropdown of folders in the ui instead?

**Comment by mmatuska** [comment]
If we remove the memory and CPUs parameters, it speeds up a lot, I am preparing a new pull request that takes care of an unset uuid

**Comment by mmatuska** [comment]
The question remains if we still want to be able to search the VMs in the filter, othewise we could try to split into folders. Folders need to be accessed as objects, otherwise you have to search for the folder parsing names of all folders and following a path. And of course via UI we cannot pass objects, or do we?

**Pull request closed without merging**